build-tests/api-extractor-scenarios/tsconfig.json

```diff
@@ -10,7 +10,14 @@
     "strictNullChecks": true,
     "types": ["node", "jest"],
     "lib": ["es5", "scripthost", "es2015.collection", "es2015.promise", "es2015.iterable", "dom"],
-    "outDir": "lib"
+    "outDir": "lib",
+
+    // These properties are only used by the pathMappings/ test scenario.
+    "baseUrl": ".",
+    "paths": {
+      "path-mapping": ["src/apiItemKinds"],
+      "wildcard-path-mapping/*": ["src/*"]
```

Comment:
The TypeScript documentation [gives an example](https://www.typescriptlang.org/tsconfig#paths) where a path is mapped to `node_modules/jquery/dist/jquery`, which probably should be considered to be external.

Could we support that easily?  Or if not, should we report a warning if we encounter `node_modules` in a path?

Reply:
From my testing in the rushstack repo, there are a few different ways to import external NPM packages, as it seems like TS has some special logic for resolving code in `node_modules`. Here's how API Extractor currently handles them:

1. `export {Foo} from 'foo'`: TS is able to resolve this to the `foo` NPM package. This is considered an external package by AE. No doc model node is created for `Foo`.
1. `export {Foo} from 'node_modules/foo'`: TS is also able to resolve this. This is also considered an external package by AE, and no doc model node is created for `Foo`.
1. `export {Foo} from '../../node_modules/foo'`: This is the explicit relative path to the `foo` package. This is considered an **internal package** by AE, and a doc model node is created for `Foo`.

When a path is mapped to `node_modules`, it's always _relative_ to the `baseUrl`. So my understanding is that any resolved path mappings are ultimately relative paths and should be interpreted by AE as such. This means that any  path mappings to `node_modules` should fall into case 3 (i.e. explicit relative path to the `foo` package), and thus should be considered part of the internal package. So maybe the PR as-is aligns with AE's current behavior?

However, it may be the case that we want AE to interpret case 3 above as an external package. I think that probably makes sense. If that's the case, then we need special logic that looks for the `node_modules` keyword and also at what path mappings resolve to (because now we can't blindly consider an import to be internal if it matches a path mapping, we need to check to see if it maps to something with `node_modules`). But I think if we want that behavior, it could be tackled in a subsequent PR?

I'll also note that in my testing, AE seems to have inconsistent behavior in creating canonical references in case 3. Sometimes the constructed `canonicalReference` looks like `"<current package>!Foo:interface"` and other times it looks like `"<external package>!Foo:interface"`. So AE may also simply have undefined behavior in case 3.

Reply:
I'm surprised `node_modules/foo` is resolved successfully at all.  That syntax definitely should be deprecated.  If API Extractor already handles it reasonably, that's okay, but if it causes any actual problems maybe we whould reject it rather than wasting any energy trying to accommodate it.

If **tsconfig.json** maps a path to `node_modules/jquery/dist/jquery`, that should definitely be considered "external" in my opinion, since it seems like an obvious (albeit misguided) way to refer to an external package. I agree that implementing this is outside the scope of this PR. However, if it's a simple change, it might be worth reporting an error for a **tsconfig.json** with such mappings. That way if we add support later, it won't be a "breaking change" for someone who relied on the broken behavior, and it creates some pressure to actually implement this properly if someone really needs it.  Up to you.

An explicit `export {Foo} from '../../node_modules/foo'` is more muddy. If we rolled that up into `import { Foo } from 'foo'` what guarantee is there that `foo` would resolve for the library consumer?  The `../../` could take you to any arbitrary folder that may be outside our **package.json**.  Thus for an `import` statement who uses a path starting with `.` or `..`, I think we're forced to treat it as a local file.  If there's a bug with Case 3, that's just a bug, and outside the scope of your PR.

Reply:
> I'm surprised `node_modules/foo` is resolved successfully at all. That syntax definitely should be deprecated. If API Extractor already handles it reasonably, that's okay, but if it causes any actual problems maybe we would reject it rather than wasting any energy trying to accommodate it.

From my local testing it seemed to work fine, but I'll confirm again tomorrow.

> Thus for an import statement who uses a path starting with `.` or `..`, I think we're forced to treat it as a local file.

I've been considering any mapped path as this exact scenario... as mapped paths are also resolved w.r.t. the `baseUrl`, which is always something like `.` or `..` or `../../something`. This is why my PR currently treats all mapped paths as local files, even a mapping to something like `node_modules/jquery/dist/jquery`. From your comment above, I know you think that mappings to things like `node_modules/jquery/dist/jquery` should be considered external, so just looking to clarify this point before moving forward.

Reply:
> The TypeScript documentation [gives an example](https://www.typescriptlang.org/tsconfig#paths) where a path is mapped to `node_modules/jquery/dist/jquery`, which probably should be considered to be external.

This is now considered to be external with the latest approach. Resolving.